pkg/plugin/builtin/v1/destination.go

```diff
@@ -0,0 +1,231 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package builtinv1
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	"github.com/conduitio/conduit-plugin/cpluginv1"
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/conduitio/conduit/pkg/foundation/log"
+	"github.com/conduitio/conduit/pkg/plugin"
+	"github.com/conduitio/conduit/pkg/plugin/builtin/v1/internal/fromplugin"
+	"github.com/conduitio/conduit/pkg/plugin/builtin/v1/internal/toplugin"
+	"github.com/conduitio/conduit/pkg/record"
+	"github.com/rs/zerolog"
+)
+
+// TODO make sure a panic in a plugin doesn't crash Conduit
+type destinationPluginAdapter struct {
+	impl cpluginv1.DestinationPlugin
+	// logger is used as the internal logger of destinationPluginAdapter.
+	logger log.CtxLogger
+	// ctxLogger is attached to the context of each call to the plugin.
+	ctxLogger zerolog.Logger
+
+	stream *destinationRunStream
+}
+
+var _ plugin.DestinationPlugin = (*destinationPluginAdapter)(nil)
+
+func newDestinationPluginAdapter(impl cpluginv1.DestinationPlugin, logger log.CtxLogger) *destinationPluginAdapter {
+	return &destinationPluginAdapter{
+		impl:      impl,
+		logger:    logger.WithComponent("builtinv1.destinationPluginAdapter"),
+		ctxLogger: logger.WithComponent("plugin").ZerologWithComponent()}
+}
+
+func (s *destinationPluginAdapter) withLogger(ctx context.Context) context.Context {
+	return s.logger.WithContext(ctx)
+}
+
+func (s *destinationPluginAdapter) Configure(ctx context.Context, cfg map[string]string) error {
+	req, err := toplugin.DestinationConfigureRequest(cfg)
+	if err != nil {
+		return err
+	}
+	s.logger.Trace(ctx).Msg("calling Configure")
+	resp, err := s.impl.Configure(s.withLogger(ctx), req)
+	if err != nil {
+		// TODO create new errors as strings, this happens when they are
+		//  transmitted through gRPC and we don't want to falsely rely on types
+		//  in built-in plugins
```

Comment:
Do you mind elaborating a bit more? I'm probably missing something, but I can't see how gRPC relates to the built-in plugins?

Reply:
We should strive to make built-in plugins work exactly the same way as standalone plugins, so given that error types are lost in gRPC communication we should actively drop the type here as well. The idea is to not start relying on some specific error type in a test, and then find out that the flow breaks if the plugin is used in standalone mode.

Reply:
Sounds great to me, and I think your comment here would be a good one to be added to the source code too.